### ISSUE TITLE
using z3 as the static library

### ISSUE THREAD

**Issue by Flians** [Z3Prover/z3#4554]
I use z3 as the static library. And I link the z3 like:
`
target_link_libraries (${PROJECT_NAME} ${CMAKE_THREAD_LIBS_INIT} ${CMAKE_SOURCE_DIR}/lib/linux/libz3.a)
`

But the following error occurred:
```
Unable to open 'pthread_mutex_lock.c': Unable to read file '/build/glibc-YYA7BZ/glibc-2.31/nptl/pthread_mutex_lock.c' 
(Error: Unable to resolve non-existing file '/build/glibc-YYA7BZ/glibc-2.31/nptl/pthread_mutex_lock.c').
```

**Comment by andrewvaughanj** [comment]
Is this when you're debugging from within Visual Studio Code:

- https://github.com/Microsoft/vscode-cpptools/issues/1123

- https://stackoverflow.com/questions/59766375/why-is-vscode-trying-to-find-library-code

?

**Comment by Flians** [comment]
@andrewvaughanj  
Yes,  you're debugging my program within Visual Studio Code. 
This is the same as my problem.
How can I solve it? I didt find the solution in your links.

**Comment by andrewvaughanj** [comment]
As far as I can tell, this is a "misleading" error message from VSC. It is effectively trying to find the source code for `pthread`, so you can step *into* the `pthread` code. One of the comments suggests it appears when they ["debug with gdb's follow-fork-mode enabled"](https://github.com/microsoft/vscode-cpptools/issues/1123#issuecomment-536737115).

Anyway, this issue is not Z3 related, but related to your IDE. If you debugged your application natively (e.g., using `gdb` directly), I imagine you wouldn't see this issue.

**Comment by Flians** [comment]
@andrewvaughanj 
I directly compile my program in the terminal using:
`
target_link_libraries (${PROJECT_NAME} ${CMAKE_SOURCE_DIR}/lib/linux/libz3.a)
`
And the following error occurred:
```
/usr/bin/ld: ../lib/linux/libz3.a(scoped_timer.o): in function `scoped_timer::scoped_timer(unsigned int, event_handler*)':
scoped_timer.cpp:(.text+0xe3): undefined reference to `pthread_create'
/usr/bin/ld: ../lib/linux/libz3.a(tactical.o): in function `par_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&)':
tactical.cpp:(.text._ZN12par_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE[_ZN12par_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE]+0x4e2): undefined reference to `pthread_create'
/usr/bin/ld: ../lib/linux/libz3.a(tactical.o): in function `par_and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&)':
tactical.cpp:(.text._ZN21par_and_then_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE[_ZN21par_and_then_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE]+0x6cc): undefined reference to `pthread_create'
/usr/bin/ld: ../lib/linux/libz3.a(sat_solver.o): in function `sat::solver::check_par(unsigned int, sat::literal const*)':
sat_solver.cpp:(.text+0x1296d): undefined reference to `pthread_create'
/usr/bin/ld: ../lib/linux/libz3.a(smt_parallel.o): in function `smt::parallel::operator()(ref_vector<expr, ast_manager> const&)':
smt_parallel.cpp:(.text+0x2590): undefined reference to `pthread_create'
/usr/bin/ld: ../lib/linux/libz3.a(parallel_tactic.o):parallel_tactic.cpp:(.text._ZN15parallel_tactic5solveER3refI5modelE[_ZN15parallel_tactic5solveER3refI5modelE]+0xe6): more undefined references to `pthread_create' follow
```
So I link `-lpthread`. But it still does not work in the terminal.

**Comment by andrewvaughanj** [comment]
This is a different failure; this means that you're now not linking `pthread` with CMake -- https://stackoverflow.com/a/29871891

**Comment by Flians** [comment]
@andrewvaughanj 
Yes, I have solved this bug. What I mean is that after I linked `-lpthread`, I still get an error: `Segmentation fault (core dumped). `
So I use VSCode to debug the program, and I got this error:
```
Unable to open 'pthread_mutex_lock.c': Unable to read file '/build/glibc-YYA7BZ/glibc-2.31/nptl/pthread_mutex_lock.c' 
(Error: Unable to resolve non-existing file '/build/glibc-YYA7BZ/glibc-2.31/nptl/pthread_mutex_lock.c').
```

**Comment by andrewvaughanj** [comment]
Yeah, and that message is an error from VSC (unrelated to Z3) -- have you tried using `gdb` from the command line on your application?

**Comment by Flians** [comment]
@andrewvaughanj 
Yes, I tried it.
But it still shows the error:
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
__GI___pthread_mutex_lock (mutex=0x0) at ../nptl/pthread_mutex_lock.c:67
67	../nptl/pthread_mutex_lock.c: No such file or directory.
```
I wanna know if I must install other libraries for libz3.a.
When I use the shared library of z3, my program works.

**Comment by andrewvaughanj** [comment]
Have you compiled both your application and Z3 with debug symbols? How does the backtrace look? What options did you pass to Z3's `configure` (e.g., did you pass `--single-threaded`)?

**Comment by Flians** [comment]
I first compile the z3 to get a static library using `--staticlib`.
Then I link the libz3.a to my program.

**Comment by NikolajBjorner** [comment]
Since you say you are using --staticlib it suggests you are using the python build system.
It has special provisions for linking pthreads for static library. 
We build and test --staticlib with every commit on Ubuntu18 so I would first suspect it is local issues on the platform and how you build your program.

With the CMake build system you could try with the configuration option:
- Z3_BUILD_LIBZ3_SHARED=OFF

**Comment by Flians** [comment]
@NikolajBjorner 
I have tried it on Ubuntu 20.04 LTS. 
But it still doesn't work.

**Comment by zhanghongce** [comment]
I have a different issue (but might be related to this I think) when using the z3 static library to get a static linked executable. (on Ubuntu 18.04)

The program is very simple:
```
#include <z3++.h>
#include <iostream>

int main() {
  z3::context c;
  z3::solver s(c);
  std::cout << "Hello world!" << std::endl;
  return 0;
}

```
and it is compiled in the normal way `g++ -I$(Z3_INCLUDE_PATH) main3.c -c -o main3.o`

(1) configure z3 like this: `./configure --staticlib` 

  (a) link using: `g++ main3.o -o main3 -static $(Z3_LIB_PATH)/libz3.a -lpthread` to get a static linked binary
  (b) link using: `g++ main3.o -o main3dyn $(Z3_LIB_PATH)/libz3.a -lpthread` to get a dynamic linked binary (but with libz3.a static linked into)

  There is segmentation fault for (a) before printing "Hello world!", but (b) can run normally. 

(2) configure z3 this way: `./configure --staticlib --debug` to see the trace dump

```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00000000006077d2 in __gthread_self () at /usr/include/x86_64-linux-gnu/c++/7/bits/gthr-default.h:686
#2  0x0000000000607cdf in std::this_thread::get_id () at /usr/include/c++/7/thread:343
#3  0x0000000000607c21 in __static_initialization_and_destruction_0 (__initialize_p=1, __priority=65535) at ../src/util/util.cpp:48
#4  0x0000000000607c3e in _GLOBAL__sub_I_util.cpp(void) () at ../src/util/util.cpp:172
#5  0x00000000019d78cc in __libc_csu_init ()
#6  0x00000000019d7097 in __libc_start_main ()
#7  0x000000000040104a in _start ()
(gdb) 

```

(3) configure z3 this way: `./configure --staticlib --single-threaded`

  (a) link using: `g++ main3.o -o main3 -static $(Z3LIB)/libz3.a -lpthread` to get a static linked binary
  (b) link using: `g++ main3.o -o main3dyn $(Z3LIB)/libz3.a -lpthread` to get a dynamic linked binary

  Both (a) and (b) will run succesfully.

I think I might not be the only one who encountered this issue when using z3, see this post: https://stackoverflow.com/questions/60337257/segfault-due-to-static-linking-with-ocaml-and-c-libraries

**Comment by Flians** [comment]
@zhanghongce 
Yes, my problem is the same as yours.
For the shared library, my program works normally. 
For the static library, my program could work normally with --single-threaded.

**Comment by andrewvaughanj** [comment]
Same issue as here: https://www.openwall.com/lists/musl/2015/08/23/1

Do you really need a _static_ application (at the end) vs. statically linking in `libz3.a`? You can statically link `libz3.a` _without_ building a static application at the end.

If you remove `-static` this issue goes away. It seems there might be a bug in using `-static` with `std::this_thread::get_id()`.

**Comment by zhanghongce** [comment]
Thank you Andrew for the info! I do want a static linked binary (as I don't want to solve the library dependency, not just z3 but other libraries, when shipping to others). But I think `--single-threaded ` works for me. Thanks!

And surprisingly, the code snippet compiles and runs normally even **with** `-static` on my machine.

```
$ lsb_release -a
LSB Version:	core-9.20170808ubuntu1-noarch:printing-9.20170808ubuntu1-noarch:security-9.20170808ubuntu1-noarch
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```
```
$ gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by andrewvaughanj** [comment]
Hmm, well, doesn't work for me:

<img width="752" alt="Screenshot 2020-07-07 at 15 37 23" src="https://user-images.githubusercontent.com/14277540/86797613-c6eff300-c067-11ea-9d68-a7e7c672df28.png">

I think this all depends on the order in which things are statically initialised, which may vary from Linux distribution to Linux distribution, and also between versions of GCC.

How does your backtrace look (under `gdb`) for your Z3 application that doesn't work? For me, the backtrace I get between this test application (and when I statically link `libz3.a`) is the same.

**Comment by zhanghongce** [comment]
@andrewvaughanj  Ah, interesting...

The backtrace was posted earlier in this issue.

> 
> (2) configure z3 this way: `./configure --staticlib --debug` to see the trace dump
> 
> ```
> Program received signal SIGSEGV, Segmentation fault.
> 0x0000000000000000 in ?? ()
> (gdb) bt
> #0  0x0000000000000000 in ?? ()
> #1  0x00000000006077d2 in __gthread_self () at /usr/include/x86_64-linux-gnu/c++/7/bits/gthr-default.h:686
> #2  0x0000000000607cdf in std::this_thread::get_id () at /usr/include/c++/7/thread:343
> #3  0x0000000000607c21 in __static_initialization_and_destruction_0 (__initialize_p=1, __priority=65535) at ../src/util/util.cpp:48
> #4  0x0000000000607c3e in _GLOBAL__sub_I_util.cpp(void) () at ../src/util/util.cpp:172
> #5  0x00000000019d78cc in __libc_csu_init ()
> #6  0x00000000019d7097 in __libc_start_main ()
> #7  0x000000000040104a in _start ()
> (gdb) 
> ```
>

**Comment by andrewvaughanj** [comment]
Mine from that test:

```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x000000000040375b in __gthread_self () at /usr/include/c++/8/x86_64-redhat-linux/bits/gthr-default.h:686
#2  0x00000000004037dc in std::this_thread::get_id () at /usr/include/c++/8/thread:349
#3  0x00000000004036c5 in main () at test_thread.cpp:6
```

As I mentioned, there is a bug here in static linking and `std::this_thread::get_id` -- it is not Z3 specific.

**Comment by andrewvaughanj** [comment]
Btw, if you google this, there are lots of reports of this kind of thing:

- https://stackoverflow.com/questions/10390584/force-gcc-to-static-link-e-g-pthreads-and-not-dynamic-link

- https://stackoverflow.com/questions/9002264/starting-a-stdthread-with-static-linking-causes-segmentation-fault

- https://stackoverflow.com/questions/23499656/segmentation-fault-in-stdthreadids-stdoperator

- https://stackoverflow.com/questions/7090623/c0x-thread-static-linking-problem

- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52590#c3

The [top answer from here](https://stackoverflow.com/a/10390911) has this to say:

> While linking libpthread.a into a shared library is theoretically
> possible, it is a really bad idea. The reason is that libpthread is part
> of glibc, and all parts of glibc must match exactly, or you'll see
> strange and un-explainable crashes.
> 

**tl;dr**: don't statically link and use pthreads.

**Comment by XVilka** [comment]
Another solution would be to statically link with [musl](https://musl.libc.org//) libc instead, with something like:
```sh
export CC="musl-gcc"
cmake -DCMAKE_EXE_LINKER_FLAGS="-static -Os" ..
make
```

**Comment by zhanghongce** [comment]
@XVilka Got it, thanks!
